src/middleware/logger.ts

```diff
@@ -39,3 +39,94 @@ export class DefaultLogger implements Logger {
     console.debug(...args);
   }
 }
+
+/**
+ * LogBuffer hold the args that will be passed on to the actual
+ * logger when attempting to flush.
+ */
+export class LogBuffer {
+  public readonly level: string;
+  public readonly args: LogArg[];
+
+  constructor(level: string, ...args: LogArg[]) {
+    this.level = level;
+    this.args = args;
+  }
+}
+
+/**
+ * ProxyLogger attempts to temporarily hold user's logs
+ * during the function run, and attempt to flush it through
+ * the provided logger when the function returns.
+ *
+ * The expected usage of this class is,
+ * 1. store the log attempt with its arguments and level
+ * 2. if function hits a step that has been memorized, clear the _buffer
+ * 3. on function return, flush all buffers
+ *
+ * And it should be invisible to the user as much as possible.
+ */
+export class ProxyLogger implements Logger {
+  private readonly _logger: Logger;
```

Comment:
Should we use real private fields instead of the semantic `_` prefix?

```suggestion
  readonly #logger: Logger;
```

Reply:
Learned something new again! :D
Will update